db/keyval/filesystem/reader/reader.go

```diff
@@ -0,0 +1,189 @@
+	IsValid(ev fsnotify.Event) (bool, error)
+}
+
+// Reader implements file system API
+type Reader struct{}
+
+// File data represents data structure of files used for configuration
+type File struct {
+	Path string
+	Data []FileEntry `json:"data"`
+}
+
+// FileEntry is data entry - single record of key-value, where key is defined as string, and value is modelled as raw message
+// (rest of the json file under the "value").
+type FileEntry struct {
+	Key   string          `json:"key"`
+	Value json.RawMessage `json:"value"`
+}
+
+// CompareTo compares file with key-value set - new, modified and deleted entries. Result is against the parameter.
+func (f *File) CompareTo(dataSet map[string][]byte) (changed, removed []FileEntry) {
+	for key, value := range dataSet {
+		var found bool
+		for _, fData := range f.Data {
+			if fData.Key == key {
+				found = true
+				if bytes.Compare(fData.Value, value) != 0 {
+					changed = append(changed, fData)
+					break
+				}
+			}
+		}
+		if !found {
+			removed = append(removed, FileEntry{key, value})
+		}
+	}
+	for _, fData := range f.Data {
+		var found bool
+		for key := range dataSet {
+			if fData.Key == key {
+				found = true
+				break
+			}
+		}
+		if !found {
+			changed = append(changed, fData)
+		}
+	}
+
+	return
+}
+
+// PathExists returns true if provided path exists, false otherwise
+func (r *Reader) PathExists(path string) bool {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+// IsValid validates file before reading, whether it is not a temporary file
+func (r *Reader) IsValid(ev fsnotify.Event) (bool, error) {
+	// Silently skip empty event
+	if ev.Name == "" {
+		return false, nil
+	}
+	// Silently skip backup files ("~" at the end)
+	if strings.Compare(ev.Name[len(ev.Name)-1:], "~") == 0 {
+		return false, nil
+	}
+	// Validate file
+	path := strings.Split(ev.Name, "/")
+	if len(path) == 0 {
+		return false, fmt.Errorf("error validating file path: invalid input %s", ev.Name)
+	}
+	fileName := path[len(path)-1]
+	parts := strings.Split(fileName, ".")
+	if len(parts) == 0 {
+		return false, fmt.Errorf("error validating file in path %s: invalid format %s", ev.Name, fileName)
+	}
+	// Silently skip numeric files without extension (temporary files with PID of the editor)
+	if len(parts) == 1 && r.isNumeric(parts[0]) {
+		return false, nil
+	}
+	// Check file extension, skip
+	fileExtension := parts[len(parts)-1]
+	match, err := regexp.MatchString("sw.", fileExtension)
+	if err != nil {
+		return false, fmt.Errorf("error validating file %s: regex error: %v", ev.Name, err)
+	}
+	if match {
+		return false, nil
+	}
+	return true, nil
+}
+
+// ProcessFile reads file from filesystem and un-marshalls it to the required proto structure
+func (r *Reader) ProcessFile(path string) (File, error) {
+	dataSet := File{}
+	fileData, err := ioutil.ReadFile(path)
+	if err != nil {
+		return dataSet, fmt.Errorf("failed to read file from path %s: %v", path, err)
```

Comment:
These functions should rather return pointer to a `File`, so we can return nil when error occurs and avoid returning something when not needed.